contracts/src/TeleporterTokenDestination.sol

```diff
@@ -0,0 +1,161 @@
+// (c) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import {TeleporterMessageInput, TeleporterFeeInfo} from "@teleporter/ITeleporterMessenger.sol";
+import {TeleporterOwnerUpgradeable} from "@teleporter/upgrades/TeleporterOwnerUpgradeable.sol";
+import {ITeleporterTokenBridge} from "./interfaces/ITeleporterTokenBridge.sol";
+import {IWarpMessenger} from
+    "@avalabs/subnet-evm-contracts@1.2.0/contracts/interfaces/IWarpMessenger.sol";
+import {IERC20} from "@openzeppelin/contracts@4.8.1/token/ERC20/IERC20.sol";
+
+/**
+ * THIS IS AN EXAMPLE CONTRACT THAT USES UN-AUDITED CODE.
+ * DO NOT USE THIS CODE IN PRODUCTION.
+ */
+
+/**
+ * @dev Abstract contract for a Teleporter token bridge that receives tokens from a {TeleporterTokenSource} in exchange for the tokens of this token bridge instance.
+ */
+abstract contract TeleporterTokenDestination is
+    ITeleporterTokenBridge,
+    TeleporterOwnerUpgradeable
+{
+    // Warp precompile used for sending and receiving Warp messages.
+    address public constant WARP_PRECOMPILE_ADDRESS = 0x0200000000000000000000000000000000000005;
+
+    // The blockchain ID of the chain this contract is deployed on.
+    bytes32 public immutable blockchainID;
+
+    // The blockchain ID of the source chain this contract receives tokens from.
+    bytes32 public immutable sourceBlockchainID;
+    // The address of the source token bridge instance this contract receives tokens from.
+    address public immutable tokenSourceAddress;
+    // The ERC20 token this contract uses to pay for Teleporter fees.
+    IERC20 public immutable feeToken;
+
+    /**
+     * @dev Initializes this destination token bridge instance to receive
+     * tokens from the specified source blockchain and token bridge instance.
+     */
+    constructor(
+        address teleporterRegistryAddress,
+        address teleporterManager,
+        bytes32 sourceBlockchainID_,
+        address tokenSourceAddress_,
+        address feeToken_
+    ) TeleporterOwnerUpgradeable(teleporterRegistryAddress, teleporterManager) {
+        sourceBlockchainID = sourceBlockchainID_;
+        tokenSourceAddress = tokenSourceAddress_;
+        // TODO: figure out if NativeTokenDestination passes in token or not.
+        // NativeTokenDestination will pass in erc20 token it deposits native tokens to pay for fees.
+        // ERC20Destination will pass in the erc20 token it's bridging.
+        feeToken = IERC20(feeToken_);
+        blockchainID = IWarpMessenger(WARP_PRECOMPILE_ADDRESS).getBlockchainID();
+    }
+
+    /**
+     * @dev Sends tokens transferred to this contract to the destination token bridge instance.
```

Comment:
```suggestion
     * @notice Sends tokens to the specified destination token bridge instance.
     * @dev Burns the specified token amount on this chain and sends a Teleporter message to the
     * token source contract to release the source token or route them to the final destination.
```

Not 100% on the exact wording, but think splitting between @notice (what does it do) and @dev (how does it do it) makes sense here